From: Ivan Kokshaysky <ink@jurassic.park.msu.ru>
Subject: Re: 2.6.13: Crash in Yenta initialization

On Tue, Sep 13, 2005 at 08:30:53AM +0200, Olaf Hering wrote:
> The reporter has updated the bug.
> 
> https://bugzilla.novell.com/attachment.cgi?id=49717
> https://bugzilla.novell.com/attachment.cgi?id=49715
> https://bugzilla.novell.com/attachment.cgi?id=49716

Thanks, that helped.

The reason was that Acer BIOS left only _two_ bus numbers available
for _three_ cardbus controllers, so pci_scan_bridge() assigns both
numbers to the first controller and leaves two other ones uninitialized.

Please try the patch here, but note that you'll apparently have only
one cardbus slot working. If you want more, use "pci=assign-busses"
boot option.

Ivan.

Signed-off-by: Olaf Hering <olh@suse.de>

--- 2.6.14-rc1/drivers/pcmcia/yenta_socket.c	Tue Sep 13 14:16:34 2005
+++ linux/drivers/pcmcia/yenta_socket.c	Tue Sep 13 14:40:40 2005
@@ -1045,7 +1045,18 @@ static int __devinit yenta_probe (struct
 {
 	struct yenta_socket *socket;
 	int ret;
-	
+
+	/*
+	 * If we failed to assign proper bus numbers for this cardbus
+	 * controller during PCI probe, its subordinate pci_bus is NULL.
+	 * Bail out if so.
+	 */
+	if (!dev->subordinate) {
+		printk(KERN_ERR "Yenta: no bus associated with %s!\n",
+			pci_name(dev));
+		return -ENODEV;
+	}
+
 	socket = kmalloc(sizeof(struct yenta_socket), GFP_KERNEL);
 	if (!socket)
 		return -ENOMEM;

